src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs

```diff
@@ -0,0 +1,99 @@
+{
+    /// <summary>
+    /// Represents a base class for file based <see cref="IConfigurationSource"/>.
+    /// </summary>
+    /// <typeparam name="TProvider">The concrete FileConfigurationProvider to create.</typeparam>
+    public class FileConfigurationSource<TProvider> : FileConfigurationSource where TProvider : FileConfigurationProvider, new()
+    {
+        public override IConfigurationProvider Build(IConfigurationBuilder builder)
+        {
+            var provider = new TProvider();
+            InitializeProvider(builder, provider);
+            return provider;
+        }
+    }
+
+    /// <summary>
+    /// Represents a base class for file based <see cref="IConfigurationSource"/>.
+    /// </summary>
+    public abstract class FileConfigurationSource : IConfigurationSource
+    {
+        /// <summary>
+        /// Used to access the contents of the file.
+        /// </summary>
+        public IFileProvider FileProvider { get; set; }
+
+        /// <summary>
+        /// The path to the file.
+        /// </summary>
+        public string Path { get; set; }
+
+        /// <summary>
+        /// Determines if loading the file is optional.
+        /// </summary>
+        public bool Optional { get; set; }
+
+        /// <summary>
+        /// Determines whether the source will be loaded if the underlying file changes.
+        /// </summary>
+        public bool ReloadOnChange { get; set; }
+
+        public virtual IFileInfo GetFileInfo()
+        {
+            return FileProvider?.GetFileInfo(Path);
+        }
+
+        private static IFileProvider BuildDefaultFileProvider()
+        {
+#if NET451
+            var stringBasePath = AppDomain.CurrentDomain.GetData("APP_CONTEXT_BASE_DIRECTORY") as string ??
+                AppDomain.CurrentDomain.BaseDirectory ??
+                string.Empty;
+
+            return new PhysicalFileProvider(stringBasePath);
+#else
+            return new PhysicalFileProvider(AppContext.BaseDirectory ?? string.Empty);
+#endif
+        }
+
+        /// <summary>
+        /// Applies the defaults specified from the builder.
+        /// </summary>
+        /// <param name="builder"></param>
+        public virtual void ApplyFileSourceDefaults(IConfigurationBuilder builder)
+        {
+            var defaults = builder.GetFileSourceDefaults();
+            FileProvider = defaults.FileProvider ?? BuildDefaultFileProvider();
+            Optional = defaults.Optional;
+            ReloadOnChange = defaults.ReloadOnChange;
+        }
+
+        /// <summary>
+        /// Initializes the file based provider
+        /// </summary>
+        /// <param name="builder">The builder for the configuration being built.</param>
+        /// <param name="provider">The provider to initialize.</param>
+        public virtual void InitializeProvider(IConfigurationBuilder builder, FileConfigurationProvider provider)
+        {
+            ApplyFileSourceDefaults(builder);
+            provider.File = GetFileInfo();
```

Comment:
Provider should have IFileProvider and path properties, not a File property. You shouldn't close over and keep alive the same IFileInfo.